migrations/20231129172339_job_queue_table.sql

```diff
@@ -0,0 +1,16 @@
+CREATE TYPE job_status AS ENUM(
+    'available',
+    'completed',
+    'failed',
+    'running'
+);
+
+CREATE TABLE job_queue(
```

Comment:
How are we tracking next attempt timing (i.e. exponential backoff)?

Reply:
We are not, at least not yet. I'm thinking of introducing a `retryable_at` timestamp field:
* When the job is first inserted, this value is `NULL`, and it's ignored by dequeue.
* When a job fails and has to be retried this can be updated to a time in the future.
  * The time can be decided by a retry policy (i.e. exponential backoff).
* Any dequeue operations will ignore jobs with retryable_at in the future.

I think, we will also need to hold a lock beyond the call to `dequeue` as the worker won't know if it will fail as it's dequeueing the job, it only does so once it's done. So, by opening a transaction and holding a lock while the worker works, all concurrent dequeues will skip this row. Not exactly sure how this implementation will look like, working through some ideas...

Any thoughts?

Reply:
> I think, we will also need to hold a lock beyond the call to dequeue as the worker won't know if it will fail as it's dequeueing the job, it only does so once it's done. So, by opening a transaction and holding a lock while the worker works, all concurrent dequeues will skip this row. 

This was my understanding as well

Reply:
I went with `scheduled_at` for our retry implementation, just in case we ever want to support scheduling webhooks, and because the name still works with normal retrying.